### PR TITLE
feat(playground): playground backend v2

### PR DESCRIPTION
**Pull request by fiji-flo**
## Summary

This moves the playground to using a back-end and SearchParams instead of the iframe based runner.

Notes:
- The `state` search param is the compressed and base64 encoded content.
- For displaying the preview in full screen we require the playground subdomain to be the hash of the compress code it will display.
- Adding `#livesample_fullscreen=${id}` to an MDN url will redirect to the livesample with `id` in fullscreen. (This will be used for the `LiveSampleLink` macro.
- `Clear-Site-Data` is now  set to `"cache", "cookies", "storage"` instead of `"*"` to not clear the execution contexts.

### Problem

Some live samples never worked, due synthetic load events and the playground could not display a preview outside an iframe.

### Solution

Move to a back-end based approach.

---

Deployed to testing.

### CODE REVIEW

File: libs/play/index.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+ * @property {string} js
+ * @property {string} [src]
+ */
+
+/**
+ * @param {IncomingMessage | null} _proxyRes
+ * @param {IncomingMessage} _req
+ * @param {ServerResponse<IncomingMessage>} res
+ */
+export function withRunnerResponseHeaders(_proxyRes, _req, res) {
```
Comment (LeoMcA) [inline libs/play/index.js:24]: As it's only used below, we could change this to something like:

```suggestion
function setRunnerResponseHeaders(res) {
```

File: libs/play/index.js [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    const urlWithCode = new URL(url);
+    urlWithCode.searchParams.set("code", rand);
+    return res
+      .status(200)
+      .send(
+        renderWarning(
+          json,
+          `${urlWithCode.pathname}${urlWithCode.search}`,
+          url.search
+        )
```
[the remark this answers is not in the document]
Comment (fiji-flo) [reply libs/play/index.js:355]: The `search` property is percent-encoded so this is fine.

**Review by LeoMcA — COMMENTED** [review]
Waiting for the staticMiddleware stuff to be fixed to test locally and fully approve, but looking good so far - just one nit, and a couple of outstanding comments about SHA-1 and updatePlayIframe from before

File: kumascript/macros/EmbedLiveSample.ejs [hunk 3]
```diff
@@ -97,7 +97,7 @@ if (hasScreenshot) {
 %> title="<%= title %> sample" id="frame_<%= id %>"<%
 if (width) { %> width="<%= width %>"<% }
 if (height) { %> height="<%= height %>"<% }
-%> src="<%- url %>"<%
+%> src="about:blank" "data-live-path="<%=env.url%>" data-live-id="<%=id%>"<%
```
[the remark this answers is not in the document]
Comment (LeoMcA) [reply kumascript/macros/EmbedLiveSample.ejs:100]: Is the shift from `url` to `env.url` intentional? There seems to be a lot of logic going into the `url` variable which is no longer occurring

File: client/src/playground/utils.ts [hunk 4]
```diff
@@ -57,13 +57,14 @@ export async function initPlayIframe(
...
-    ? PLAYGROUND_BASE_HOST
-    : `${hash}.${PLAYGROUND_BASE_HOST}`;
+  const baseUrl = PLAYGROUND_BASE_HOST.startsWith("localhost")
+    ? `http://${PLAYGROUND_BASE_HOST}`
+    : `https://${hash}.${PLAYGROUND_BASE_HOST}`;
   // We're modifying most parts but keep path, protocoll, port and hash.
-  const url = new URL(iframe.src);
-  url.host = host;
+  const url = new URL(baseUrl);
+  url.pathname = `${path || ""}${path?.endsWith("/") ? "" : "/"}runner.html`;
```
Comment (LeoMcA) [inline client/src/playground/utils.ts]: Comment probably needs changing now.

Not a huge fan of the hard-coded `http` for `localhost`, could we do something like:

```js
  const host = PLAYGROUND_BASE_HOST.startsWith("localhost")
    ? `${PLAYGROUND_BASE_HOST}`
    : `${hash}.${PLAYGROUND_BASE_HOST}`;
  const url = new URL(path, window.location.origin);
  url.host = host
```

**Review by LeoMcA — COMMENTED** [review]
I misled you over the `he` types, forgot typescript is annoying with es modules, the change to `import * as he` wasn't actually working: I pushed a fix

**Review by LeoMcA — APPROVED** [review]
lgtm, let's just get the tests to pass! 🚀

**Pull request merged**